packages/venia-concept/src/components/ProductImageCarousel/carousel.js

```diff
@@ -4,34 +4,62 @@ import PropTypes from 'prop-types';
 import classify from 'src/classify';
 import ThumbnailList from './thumbnailList';
 import defaultClasses from './carousel.css';
-
-const uri =
-    'iVBORw0KGgoAAAANSUhEUgAAAAQAAAAFCAQAAADIpIVQAAAADklEQVR42mNkgAJGIhgAALQABsHyMOcAAAAASUVORK5CYII=';
+import { makeProductMediaPath } from 'src/util/makeMediaPath';
+import { transparentPlaceholder } from 'src/shared/images';
 
 class Carousel extends Component {
     static propTypes = {
         classes: PropTypes.shape({
-            currentImage: PropTypes.string,
+            currentImageFile: PropTypes.string,
             root: PropTypes.string
         }),
         images: PropTypes.arrayOf(
             PropTypes.shape({
-                id: PropTypes.string
+                label: PropTypes.string,
+                position: PropTypes.number.isRequired,
+                disabled: PropTypes.bool,
+                file: PropTypes.string.isRequired
             })
-        )
+        ).isRequired
     };
 
+    shouldComponentUpdate({ images }) {
+        return (
```

Comment:
I don't think this should be a `shouldComponentUpdate`. This actually will run (in most cases) just as frequently as `render`. So now, if nothing has changed, we're looping over all the images _3 times_ (the filter in `render`, the sort in `render`, and the `some` in here).

[The recommended approach](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization) for possibly-expensive derivations from state in `render` is to use memoization. The better alternative is `PureComponent`, but we're writing these components for third parties, so we can't guarantee they'll always pass in an array with a new identity.

/cc @jimbo to correct me if I'm wrong about anything here

Reply:
Memoizing!